aten/src/ATen/native/TensorShape.cpp

```diff
@@ -27,6 +28,147 @@
 #include <vector>
 
 namespace at {
+namespace meta {
+static void cat_check_no_zero_dim(ITensorList tensors) {
+  for (const auto i : c10::irange(tensors.size())) {
+    auto& t = tensors[i];
+    TORCH_CHECK(
+        t.dim() > 0,
+        "zero-dimensional tensor (at position ", i, ") cannot be concatenated");
+  }
+}
+
+static c10::MemoryFormat cat_compute_output_memory_format(ITensorList inputs) {
+  c10::optional<c10::MemoryFormat> format = c10::nullopt;
+  for (auto& t : inputs) {
+    auto f = t.suggest_memory_format();
+    if (!format.has_value()) {
+      format = f;
+      continue;
+    }
+    if (format.value() == f) {
+      continue;
+    }
+    bool contiguous =
+        (format.value() == c10::MemoryFormat::Contiguous ||
+         f == c10::MemoryFormat::Contiguous || format.value() != f);
+    if (contiguous) {
+      return c10::MemoryFormat::Contiguous;
+    }
+  }
+  return format.value();
+}
+
+TORCH_PRECOMPUTE_META_FUNC(cat)(ITensorList tensors, int64_t dim) {
+  // previously, size [0] tensors were the only possible empty tensors; thus, it wasn't possible
+  // to cat empty tensors unless all the other tensors were 1-dimensional, so we allowed these tensors
+  // to be "skipped".  We maintain this behavior for backwards compatibility, but only for this specific
+  // size (i.e. other empty sizes are not skipped).
+  cat_check_no_zero_dim(tensors);
+  dim = at::legacy_cat_wrap_dim(dim, tensors);
+
+  TORCH_CHECK(
+      tensors.size() > 0, "torch.cat(): expected a non-empty list of Tensors");
+
+  // Look for the first valid tensor.
+  size_t valid = tensors.size();
+  for (const auto i : c10::irange(tensors.size())) {
+    if (!at::native::cat_should_skip_tensor(tensors[i])) {
+      valid = i;
+      break;
+    }
+  }
+
+  bool all_contiguous = true;
+  bool all_same_dtype = true;
+  bool all_same_sizes_and_stride = true;
+  auto memory_format = cat_compute_output_memory_format(tensors);
```

Comment:
CPU and CUDA kernels computed the memory format in different ways. Here, I adopted how CUDA used to do it. Not sure if this is the best way to go at it. Any thoughts?

Reply:
cc @ngimel for another opinion. I found the PR that [added support for cuda](https://github.com/pytorch/pytorch/pull/34526), and a [bit later for cpu](https://github.com/pytorch/pytorch/pull/39448). There's no explicit mention of why they differ, so, going with the way cuda does it now seems reasonable to me.

Reply:
@mruberry helped point me to some discussion about this: https://github.com/pytorch/pytorch/pull/62560#issuecomment-905923630.

It sounds like making cuda's behavior the general behavior is the right move in this PR. We should also add a test (if one doesn't exist already), that confirms that memory_format behavior is the same across cpu and cuda.

@mruberry also pointed out another thing we need to fix in this PR: see https://github.com/pytorch/pytorch/issues/64709. Apparently, there parts of the codebase that call `cat` and expect `cat(out=...)` to resize the output tensor. That'll start printing warnings now that `cat` is structured. We should fix all of the places in our codebase that call `cat(out=...)` and expect the resize to happen, that way users don't start seeing warnings that they can't fix.

Reply:
Existing cat memory_format test is here: https://github.com/pytorch/pytorch/blob/1d269e8c15da8bd4a3bfcfbf44ae27ec36e9735b/test/test_tensor_creation_ops.py#L721

Reply:
Thanks @ysiraichi, indeed the CUDA and CPU memory format calculation is inconsistent. I opened issue #63998 (talking exactly about this) earlier and wanted to work on a fix for that but never got around that because of some new responsibilities until earlier this week. I just came across this PR by chance and I see you have already incorporated the change.

So, with this PR, #63998 should also be fixed.

@bdhirsh when adding the existing cat memory format test, I kept the different behaviour of memory formats in mind. It will now need slight modification with your update making CPU and CUDA consistent, which should be quite straightforward.